### ISSUE TITLE
Add `Flow.submit` interface for subflows

### ISSUE THREAD

**Issue by zanieb** [PrefectHQ/prefect#6689]
### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar request and didn't find it.
- [X] I searched the Prefect documentation for this feature.

### Prefect Version

2.x

### Describe the current behavior

Currently, subflows can only be run in the main process and only tasks can be submitted to other infrastructure and run in the background.

### Describe the proposed behavior

Flows should be submittable to run in the background. Calling `Flow.submit` should return a future that can be used to wait for the flow run's completion.

As a stretch and perhaps future goal: Flows should be submittable to external infrastructure without a deployment.

All of the features available for flow calls should be available for submission. We may limit this in the first iteration to get the feature in user's hands faster.

### Example Use

```python
from prefect import flow

@flow
def bar():
   return 1

@flow
def foo():
    future = bar.submit()  # run in a worker thread
    x = future.result()
```
    

As a stretch goal:

```python
from prefect import flow

@flow(infrastructure=KubernetesJob(...))
def bar():
   return 1

@flow
def foo():
    future = bar.submit()
    x = future.result()
```
    

### Additional context

An extension of https://github.com/PrefectHQ/prefect/issues/6688

This feature will require expert changes to Prefect internals.

For in a background thread:
- Futures need to be updated to support flows
- Waiting for flow run completion from a future needs to be performant
- It needs to be clear where the flow is being submitted to

For external infrastructure:
- Infrastructure declaration needs to be added at the flow decorator level
- Flow result retrieval from a future will require design
   - We may need to require result handling to be configured to use this feature
   - We may be able to pass results back via cloudpickle and stdout or a mounted file
- For container based infrastructure, we will need to determine what image the flow should run in
   - Since the user is calling it from the current flow we may be able to use the parent's image in some cases
- We need to determine how to get the flow to the infrastructure, i.e. serialization

**Comment by madkinsz** [comment]
@polivbr This interface would likely avoid the use of agents, I believe your use-case is addressed by `run_deployment` (#7047).

**Comment by anna-geller** [comment]
cc @anticorrelator - tagging you because I know you wanted to write a raft doc about it

use cases we should consider: 
- being able to run downstream flows or tasks even when one subflow fails
- being able to submit several child flows for concurrent execution

```python
from datetime import date
from prefect import flow
from flows.transformation.jaffle_shop.dbt_run_from_manifest import dbt_jaffle_shop
from flows.ingestion.ingest_jaffle_shop import raw_data_jaffle_shop
from flows.analytics.dashboards import dashboards
from flows.ml.sales_forecast import sales_forecast, run_critical_work


@flow
def jaffle_shop_ingest_transform(
    start_date: date = date(2022, 11, 1),  # parametrized for backfills
    end_date: date = date.today(),
    dataset_size: int = 10_000,
):
    raw_data_jaffle_shop(start_date, end_date, dataset_size)
    dbt_jaffle_shop()
    dashboards()
    sales_forecast()
    run_critical_work()
````

I want to run the `dashboards` and `sales_forecast` subflows concurrently and I want to run `run_critical_work` only when those two subflows are complete, even if they fail. Currently, if any subflow run fails, `run_critical_work` cannot be executed. This would be possible with the `allow_failure` annotation added in [PR 7120](https://github.com/PrefectHQ/prefect/pull/7120) but it's only applicable to tasks.

**Comment by anna-geller** [comment]
another user reporting a similar request: https://discourse.prefect.io/t/running-subflows-in-parallel-with-downstream-dependency/1903/2 

for this user, this is something that makes migration difficult

**Comment by madkinsz** [comment]
@anna-geller 

> Currently, if any subflow run fails, run_critical_work cannot be executed. 

If you run your subflows with `return_state=True` failed states will not be raised and a downstream can still be run.

I agree we should definitely mirror all the task futures here.

**Comment by kpweiler** [comment]
If/When this implemented - will it require the use of `async def` on the subflow functions (and then, transitively, the task functions)?

**Comment by madkinsz** [comment]
@kpweiler no, this would be available as a sync and async interface.

**Comment by jmesterh** [comment]
FWIW I am trying to build a POC of a 10,000 table ETL using Prefect and just ran into this issue. The first attempt looked like this (using task tags to set concurrency limits):
```python
@task
def extract()...
@task
def transform()...
@task
def load()...

@flow
def main():
  loads = []
  # ETL 10,000 tables in parallel 
  for table in tables:
    extracted = extract.submit(table)
    transform = transform.submit(extracted)
    loads.append(load.submit(transform))
  for load in loads:
    result = load.result()
    print("load complete for "+result)
```
This works, but it creates 30,000 tasks waiting for their dependencies to complete. What I would like to do is call each task sequentially (per table) while processing the tables in parallel:
```python
@task
def extract()...
@task
def transform()...
@task
def load()...

@flow
def etl(table):
    return load(transform(extract(table)))

@flow
def main()
  loads = []
  # ETL 10,000 tables in parallel
  for table in tables:
    loads.append(etl.submit(table))
  for load in loads:
    result = load.result()
    print("load complete for "+result)
```
I see there is an [async workaround](https://discourse.prefect.io/t/how-can-i-run-multiple-subflows-or-child-flows-in-parallel/96/10), but I can't use async as there is no async driver for Oracle or MSSQL in SQLAlchemy. There is also the issue of [not being able to run multiple instances of the same sub-flow](https://github.com/PrefectHQ/prefect/issues/5853). If these could be fixed so that the example above works that would be awesome.

**Comment by rsampaths16** [comment]
> only tasks can be submitted to external infrastructure and run in the background

Hi @madkinsz can you please provide example for this. I tried looking in the docs but couldn't find a way to do this. Thanks.

**Comment by daniel-oh-sunpower** [comment]
Just a gentle question: is this feature still in active development? If not, anyone else interested in seeing this implemented? (please up-vote if so!)

**Comment by nicolasiltis** [comment]
I'm very interested in seeing this feature implemented !! It would be awesome for the orchestration.
Just adding an idea here and open the discussion : 
We have TaskRunner, why can't we have FlowRunner in the same way ? (SequentialFlowRunner[default], ConcurrentFlowRunner, DaskFlowRunner, RayFlowRunner, ...)

**Comment by ricosaurus** [comment]
Chiming in that it would be great to know if this is still under development -- for me it is currently a major obstacle for using prefect.  I had naively assumed I could run subflows in parallel (without async) and am considering next steps.

**Comment by williamjamir** [comment]
I'm excited about this feature too! It would make the Prefect API more consistent, in my view, since flows and tasks will share common functionalities; big win for simplicity.

My user case is because I want to flow with different parameters, like hyperparameters. 
It can make experiment executions more simple :)

**Comment by wakatara** [comment]
Have to echo @ricosaurus here, I had assumed being able to concurrently run a subflow (mine just takes a list of files to be processes and then I wanted it to process up to the concurrency limit the "subflow" it submits them to would have been fairly trivial, so right now this is completely blocking me migratig over to Prefect from Airflow (which I'd like to migrate from, but which supports concurrency in-DAGs.). 

Can we understand when we may see this? I feel Prefect is the superior product and like the way it works much betetr (also feel you can modularly code),but no parallel execution of subflows is kinda a deal killer.

**Comment by snowdrop4** [comment]
The [docs on flows](https://docs.prefect.io/latest/concepts/flows/#composing-flows), I feel, are quite misleading. They make it seem like parallel execution of sub-flows is already supported:

> Subflow runs behave like normal flow runs. There is a full representation of the flow run in the backend as if it had been called separately

The above makes it sound like they behave like normal flows (literally). But really, they aren't comparable, as you can launch multiple flows and have them run in parallel, but this isn't the case for sub-flows.

> Subflows will block execution of the parent flow until completion. However, asynchronous subflows can be run in parallel by using [AnyIO task groups](https://anyio.readthedocs.io/en/stable/tasks.html) or [asyncio.gather](https://docs.python.org/3/library/asyncio-task.html#id6).

And then the above makes it sound like all you need to do is use asyncio.gather, and then the sub-flows will run in parallel. This also isn't true.

**Comment by kevingrismore** [comment]
@snowdrop4 I think this is a fair point. Docs will be updated in the next release to word that section slightly differently. In-process subflow runs can be concurrent and take advantage of async execution, but are not parallel in the way separate, simultaneous deployment runs are.

**Comment by snowdrop4** [comment]
Thank you :pray:

**Comment by thomasfrederikhoeck** [comment]
Hi @kevingrismore . Is this something which is actively worked on?

**Comment by daviddwlee84** [comment]
In my scenario, I have to process different date/item's data.
Each day/item is independent and can be run parallel.
And a single-day task is too complex to be a single task. (exists some sequential dependency, parallel calculation, cache for middle results, etc.)
In this scenario, submitting a subflow is required for running these subflows parallel.
https://github.com/PrefectHQ/prefect/issues/7322
https://github.com/PrefectHQ/prefect/discussions/12563

**Comment by glesperance** [comment]
This feature would go a long way in making prefect more intuitive to adopt. Moreover, since we can't run multiple flows at the same time one has to wrap them in dummy tasks to achieve paralellism.

**Comment by MichaelGFagan** [comment]
Adding myself to the list of people who likely spent hours trying to figure out how to run subflows concurrently before winding up here.

**Comment by cicdw** [comment]
Hey everyone - I'm happy to share that we _are_ actively working on an interface like this right now. We have an unpolished MVP working for submitting ad-hoc flow and subflow runs to kubernetes and will be expanding that to other compute once we have the pieces polished. I don't yet have a firm date that you can expect a release, but most likely we'll make some noise about an alpha version sometime in April and I will share more news on this issue when I have it.
